merkle-tree/hash-set/src/lib.rs

```diff
@@ -0,0 +1,1211 @@
+    pub fn contains(&self, value: &BigUint, sequence_number: usize) -> Result<bool, HashSetError> {
+        self.hash_set.contains(value, sequence_number)
+    }
+
+    /// Marks the given element with a given sequence number.
+    pub fn mark_with_sequence_number(
+        &mut self,
+        value: &BigUint,
+        sequence_number: usize,
+    ) -> Result<(), HashSetError> {
+        self.hash_set
+            .mark_with_sequence_number(value, sequence_number)
+    }
+
+    pub fn iter(&self) -> HashSetIterator<I> {
+        self.hash_set.iter()
+    }
+}
+
+impl<'a, I> Drop for HashSetZeroCopy<'a, I>
+where
+    I: Bounded
+        + CheckedAdd
+        + CheckedSub
+        + Clone
+        + Copy
+        + fmt::Display
+        + From<u8>
+        + PartialEq
+        + PartialOrd
+        + ToBigUint
+        + TryFrom<u64>
+        + TryFrom<usize>
+        + Unsigned,
+    usize: TryFrom<I>,
+    <usize as TryFrom<I>>::Error: fmt::Debug,
+{
+    fn drop(&mut self) {
+        // SAFETY: Don't do anything here! Why?
+        //
+        // * Primitive fields of `HashSet` implement `Copy`, therefore `drop()`
+        //   has no effect on them - Rust drops them when they go out of scope.
+        // * Don't drop the dynamic fields (`indices` and `values`). In
+        //   `HashSetZeroCopy`, they are backed by buffers provided by the
+        //   caller. These buffers are going to be eventually deallocated.
+        //   Performing an another `drop()` here would result double `free()`
+        //   which would result in aborting the program (either with `SIGABRT`
+        //   or `SIGSEGV`).
+    }
+}
+
+pub struct HashSetIterator<'a, I>
+where
+    I: Bounded
+        + CheckedAdd
+        + CheckedSub
+        + Clone
+        + Copy
+        + fmt::Display
+        + From<u8>
+        + PartialEq
+        + PartialOrd
+        + ToBigUint
+        + TryFrom<u64>
+        + TryFrom<usize>
+        + Unsigned,
+    usize: TryFrom<I>,
+    <usize as TryFrom<I>>::Error: fmt::Debug,
+{
+    hash_set: &'a HashSet<I>,
+    current: usize,
+}
+
+impl<'a, I> Iterator for HashSetIterator<'a, I>
+where
+    I: Bounded
+        + CheckedAdd
+        + CheckedSub
+        + Clone
+        + Copy
+        + fmt::Display
+        + From<u8>
+        + PartialEq
+        + PartialOrd
+        + ToBigUint
+        + TryFrom<u64>
+        + TryFrom<usize>
+        + Unsigned,
+    usize: TryFrom<I>,
+    <usize as TryFrom<I>>::Error: fmt::Debug,
+{
+    type Item = (usize, &'a HashSetCell);
+
+    fn next(&mut self) -> Option<Self::Item> {
+        if self.current < self.hash_set.capacity_values {
+            let element_index = self.current;
+            let element = unsafe { &*self.hash_set.values.as_ptr().add(element_index) };
+
+            self.current += 1;
+            element.as_ref().map(|element| (element_index, element))
+        } else {
+            None
+        }
+    }
+}
+
+#[cfg(test)]
+mod test {
```

Comment:
Hm there is the random test which tests this somewhat. Do you think it makes sense to add an explicit test for this?
For example:
0. hashset with capacity 1
1. insert value with sequence 10
2. try to overwrite with sequence 9 should fail
3. overwrite with 10 should succeed

good question is should 10 or 11 succeed?

Reply:
Random tests do that. But yes, I can add a manual one.

With the current implementation, 10 is going to succeed.

Reply:
Welp, nope, I double checked. 10 wouldn't succeed. 11 would succeed.

Reply:
Ok I think this should be safe.
Thinking out loud how we use it:
- in compressed pda execute compressed transaction:
1. do some checks verify proof etc...
2. append new compressed accounts as leaves
3. nullify input compressed accounts by inserting them into the hashset
-> since the append operation already happend in the transaction the sequence number we read when inserting the nullifier into the hashset is updated. Once we did root history array length state root updates the root history array has been completely overwritten. Once it is compeletly overwritten it is safe to overwrite.

Ah lol while writing this down I noticed that we don't set the number to overwrite when inserting the value into the hashset initially, instead we set it 0 when inserting it initally and mark as ready to be overwritten once it the value has been nullified in the state Merkle tree or the value has been inserted into the address Merkle tree.

[see insertion into indexed array](
https://github.com/Lightprotocol/light-protocol/blob/afed93d0bab67ef49d0a9d7c6bce0d6dab58ac6e/programs/account-compression/src/instructions/insert_into_indexed_array.rs#L61-L63)
[see actual nullify instruction](https://github.com/Lightprotocol/light-protocol/blob/afed93d0bab67ef49d0a9d7c6bce0d6dab58ac6e/programs/account-compression/src/instructions/nullify_leaves.rs#L125-L128)

It should be the same way for the address Merkle tree.